src/VisualStudio/Core/Def/Implementation/FindUsages/CodeIndexExternalReferenceItem.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.FindUsages;
+using Microsoft.CodeAnalysis.Text;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.FindUsages
+{
+    internal partial class VisualStudioFindSymbolMonikerUsagesService
+    {
+        private class CodeIndexExternalReferenceItem : ExternalReferenceItem
+        {
+            private readonly VisualStudioFindSymbolMonikerUsagesService _service;
+            private readonly JObject _resultObject;
+
+            public CodeIndexExternalReferenceItem(
+                VisualStudioFindSymbolMonikerUsagesService service,
+                DefinitionItem definition,
+                JObject resultObject,
+                string projectName,
+                string displayPath,
+                LinePositionSpan span,
+                string text) : base(definition, projectName, displayPath, span, text)
+            {
+                _service = service;
+                _resultObject = resultObject;
+            }
+
+            public override bool TryNavigateTo(bool isPreview)
+            {
+                // Cancel the navigation to any previous item the user was trying to navigate to.
+                // Then try to navigate to this. Because it's async, and we're not, just assume it
+                // will succeed.
+                var cancellationToken = _service.CancelLastNavigationAndGetNavigationToken();
+                _ = NavigateToAsync(isPreview: false, cancellationToken);
+                return true;
+            }
+
+            private async Task NavigateToAsync(bool isPreview, CancellationToken cancellationToken)
+            {
+                // No way to report any errors thrown by OpenNavigationResultInEditorAsync.
+                // So just catch and report through our watson system.
+                try
+                {
+                    await _service._codeIndexProvider!.OpenNavigationResultInEditorAsync(
+                        _resultObject, isPreview, cancellationToken).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                }
+                catch (Exception e) when (FatalError.ReportWithoutCrash(e))
+                {
+                }
+            }
```

Comment:
Want to just move the core of this into VisualStudioFindSymbolMonikerUsagesService? Then you don't need to expose a CancelLastNavigationAndGetNavigationToken since this can just live internal as an implementation detail?

Reply:
sure.

Reply:
great idea!